#bisulfite1_initialize.R
#Groves Dixon


#set working dir
setwd("~/gitreps/reciprocal_transplant_methylation/bisulfite_validation/")
source("~/gitreps/reciprocal_transplant_methylation/scripts/bisulfite_functions.R")
library(reshape)
library(plotrix)

#------------ UPLOAD THE BISULFITE DATA ------------#

#upload sample name spreadsheet
snames = read.csv('data/sample_names.csv')

#upload bismark data
d = read.table('data/all_PE_bismark_results.tsv', sep="\t", header = T) #(see Bisulfite_amplicon_data_process_Walkthrough.txt to generate)
head(snames)
head(d)


#------------ FILTER AND FORMAT THE BISULFITE DATA ------------#

#remove lines with fewer than 50 reads
sums=d$methReads + d$UnmethReads
d0=d
d=d[sums>50,]
dim(d0)
dim(d)

#swap in the sample names for the bs ids
sample = as.character(lapply(d$file, function(x) as.character(snames$sample[snames$bsName == x][1])))
d$sample = sample
d$ori = substr(sample, start=1, stop=1)
d$trans = substr(sample, start=2, stop=2)
tp = c()
for (s in sample){
	time=strsplit(s, "_")[[1]][2]
	tp=append(tp, time)
}
d$tp=tp
d$cpgSite = factor(d$cpgSite)
head(d)

#set up genotype variable
geno = c()
for (s in d$sample){
	geno0 = strsplit(s, "_")[[1]][1]
	geno1 = substr(geno0, start = 1, stop = 1)
	geno2 = substr(geno0, start = 3, stop = 5)
	genoFinal=paste(geno1, geno2, sep="")
	geno=append(geno, genoFinal)
}
d$geno=geno


#------------ REMOVE CPG CALLS THAT WERE INSIDE OF PRIMER BINDING SITES ------------#

#upload the locus names for the bisulfite tested genes
bsgenes = read.csv('../datasets/bisulfite_genes.csv') #note this file includes the 3' coordinates for the forward and reverse primers
bsgenes$locus = paste('locus', bsgenes$primerset, sep="")
rownames(bsgenes) = bsgenes$gene
head(bsgenes)
head(d)

for (l in bsgenes$locus){
	sub = d[d$locus == l,]
	psub=bsgenes[bsgenes$locus==l,]
	unique(sub$position)
	inFor = unique(sub$position[sub$position < psub$f3prime])
	inRev = unique(sub$position[sub$position > psub$r3prime])
	bad=c(inFor, inRev)
	remove = paste(paste(l, c(inFor, inRev), sep = "_"), bad, sep = "_")
	print("-------------------")
	print(paste("Removing bad CpG sites for", l))
	print('Bad Positions:')
	print(bad)
	d=d[!d$cpgSite %in% remove,]
}
#reset the levels
d$cpgSite = factor(d$cpgSite)


#subset for timepoint 2
alld = d
d=d[d$tp == 2,]
dat=d

#------------ LOOK AT DISTRIBUTION OF METHYLATION PERCENTAGES ------------#
par(mfrow=c(3,1))
hist(d$methPct, main = "All CpG Sites", xlab="Percent Methylated")
cut=8
weak=d[d$methPct < cut,]
wpct = paste(round(nrow(weak)/nrow(d)*100, digits=1), "% of CpG Sites", sep="")
strong=d[d$methPct > cut,]
spct = paste(round(nrow(strong)/nrow(d)*100, digits=1), "% of CpG Sites", sep="")
hist(weak$methPct, main = paste("Unmethylation Sites", wpct, sep="\n"), xlab="Percent Methylated")
hist(strong$methPct, main = paste("Methylated Sites", spct, sep="\n"), xlab="Percent Methylated")

#------------ CORRELATE WITH CAPTURED VS FLOWTHROUGH ------------#

#upload the capture vs flowthrough methylation scores for each gene
lnames = load('../datasets/capturedVflowthroughResults_p-1000_200.Rdata') #generated by DEseq_unbound_v_flowthrough
lnames
#format the data
mbd.dat = data.frame(res$log2FoldChange)
rownames(mbd.dat) = rownames(res)
colnames(mbd.dat) = c('mbd.score')
mbd.dat=na.omit(mbd.dat)
#clean up unneeded variables
for (i in lnames){
	remove(i)
}
head(mbd.dat)

#merge with the locus data to fix names
mdat = merge(mbd.dat, bsgenes, by = 0)
head(mdat)
dim(bsgenes)
dim(mdat)

#get mean methylation for 
x = data.frame(tapply(d$methPct, d$locus, mean))
colnames(x) = c('methPct')
x$locus = rownames(x)
gene.mns = merge(x, mdat, by = 'locus')
head(gene.mns)
dim(gene.mns)
par(mfrow=c(1,1), xpd=F)
plot(gene.mns$mbd.score~gene.mns$methPct, xlab = 'Mean % Methylation', ylab = 'MBD-score')
lm1=lm(gene.mns$mbd.score~gene.mns$methPct)
summary(lm1)
abline(lm1, col='red')
#reverse the axes
plot(gene.mns$methPct~gene.mns$mbd.score, ylab = 'Mean % Methylation', xlab = 'MBD-score')
lm1=lm(gene.mns$methPct~gene.mns$mbd.score)
summary(lm1)
abline(lm1, col='red')

#log transformed
plot(log(gene.mns$methPct,2)~gene.mns$mbd.score, ylab = expression(paste('log'[2], ' Mean % Methylation')), xlab = 'MBD-score', cex=1, axes=T)
# axis(1);axis(2)
lm1=lm(log(gene.mns$methPct,2)~gene.mns$mbd.score)
summary(lm1)
abline(lm1, col='red')



#do it Mclearth 2016 style (Statistical Rethinking)
library('rethinking')
gene.mns$logm=log(gene.mns$methPct, 2)
#set up model using map
m1 <- map(
		alist(
		logm ~ dnorm(mu, sigma),
		mu <- a + b*mbd.score,
		a~dnorm(0, 10),
		b~dnorm(.1, 2),
		sigma~dunif(0,50)
		), 
		data= gene.mns)
	

precis(m1)
mbd.seq <- seq(from=-4, to=0.5, length.out=1000)
pred.dat <- list(mbd.score=mbd.seq)
mu<-link(m1, data=pred.dat)
mu.mean = apply(mu, 2, mean)
mu.PI <- apply(mu, 2, PI, prob=.95)
sim.mbd<-sim(m1, data=pred.dat)
mbd.PI<-apply(sim.mbd, 2, PI, prob=.95)

#plot results
plot(log(gene.mns$methPct,2)~gene.mns$mbd.score, ylab = expression(paste('log'[2], ' Mean % Methylation')), xlab = 'MBD-score', cex=1, axes=T,mgp=c(2.1,1,0))
lines(mbd.seq, mu.mean, col='red')
shade(mu.PI, mbd.seq)
shade(mbd.PI, mbd.seq)


#save outputs
save(gene.mns, file="../datasets/methPct_mbd_score.Rdata")
save(bsgenes, d, file='../datasets/bisulfite_dat.Rdata')
